Source/Engine/Utilities/Extensions.cs

```diff
@@ -286,7 +286,28 @@ public static float NextFloat(this Random random, float max)
         /// <returns>A random <see cref="Quaternion"/>.</returns>
         public static Quaternion NextQuaternion(this Random random, bool randomRoll = false)
         {
-            return Quaternion.Euler(NextFloat(random, -180.0f, 180.0f), NextFloat(random, -180.0f, 180.0f), randomRoll ? NextFloat(random, -180.0f, 180.0f) : 0.0f);
+            //Based on https://msl.cs.uiuc.edu/planning/node198.html
+            float u1 = NextFloat(random);
+            float u2 = NextFloat(random);
+            float u3 = NextFloat(random);
+            float rad = 2.0f * Mathf.Pi; 
+
+            float sqrt1MinusU1 = Mathf.Sqrt(1 - u1);
+            float sqrtU1 = Mathf.Sqrt(u1);
+
+            float x = sqrt1MinusU1 * Mathf.Sin(rad * u2);
+            float y = sqrt1MinusU1 * Mathf.Cos(rad * u2);
+            float z = sqrtU1 * Mathf.Sin(rad * u3);
+            float w = sqrtU1 * Mathf.Cos(rad * u3);
+
+            Quaternion result = new Quaternion(x, y, z, w);
+
+            if (randomRoll)
```

Comment:
Should be `if (!randomRoll)` I suppose? I noticed it just now.

Reply:
Yeah. I was tired when I wrote this function.
I thought I put first to return with a roll and then without.
It's funny that we both didn't notice it at first :D